### PR TITLE
cleanup(GCS+gRPC): use protos for `ReadObject*()`

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #13910

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14032)

### CODE REVIEW

File: google/cloud/storage/async/client.cc [hunk 1]
```diff
@@ -38,6 +38,53 @@ AsyncClient::AsyncClient(Options options) {
...
+    std::int64_t limit, Options opts) {
+  auto request = google::storage::v2::ReadObjectRequest{};
+  request.set_bucket(bucket_name.FullName());
+  request.set_object(std::move(object_name));
+  request.set_read_offset(offset);
+  request.set_read_limit(limit);
+
+  return connection_->ReadObjectRange(
+      {std::move(request),
+       internal::MergeOptions(std::move(opts), connection_->options())});
```
Comment (devbww) [inline google/cloud/storage/async/client.cc:74]: Could not this forward to `ReadObjectRange(request, ...)`, to match the layout of the `ReadObject()` overloads?

File: google/cloud/storage/async/client_test.cc [hunk 2]
```diff
@@ -92,20 +92,24 @@ TEST(AsyncClient, InsertObject) {
...
+    object: "test-object"
+  )pb";
   auto mock = std::make_shared<MockAsyncConnection>();
   EXPECT_CALL(*mock, options)
       .WillRepeatedly(
           Return(Options{}.set<TestOption<0>>("O0").set<TestOption<1>>("O1")));
 
   EXPECT_CALL(*mock, ReadObject)
-      .WillOnce([](AsyncConnection::ReadObjectParams const& p) {
+      .WillOnce([&](AsyncConnection::ReadObjectParams const& p) {
```
Comment (devbww) [inline google/cloud/storage/async/client_test.cc:106]: :smile:

File: google/cloud/storage/async/connection.h [hunk 3]
```diff
@@ -85,7 +85,7 @@ class AsyncConnection {
     /// The name of the bucket and object to read. Includes optional parameters,
     /// such as pre-conditions on the read operation, or the range within the
     /// object to read.
```
Comment (devbww) [inline google/cloud/storage/async/connection.h:87]: I presume the comment still holds despite the type change.

**Review by devbww — APPROVED** [review]

**Review by scotthart — APPROVED** [review]
> Reviewed 15 of 15 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14032)* status: all files reviewed, 3 unresolved discussions (waiting on @coryan)

File: google/cloud/storage/async/client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/async/client.cc]: Done.

File: google/cloud/storage/async/connection.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/async/connection.h:87]: Yes, it does.

File: google/cloud/storage/async/client_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/async/client_test.cc:106]: I try to learn. I also fail to learn, but I try to.

**Pull request merged**